test/e2e/upgrade_management_cluster_test.go

```diff
@@ -0,0 +1,352 @@
+	By("Get the management cluster images before upgrading")
+	printImages(upgradeClusterProxy)
+
+	Logf("Generate clusterctl file")
+	pwd, err := os.Getwd()
+	Expect(err).To(BeNil())
+	Logf("PWD:", pwd)
+
+	type vars struct {
+		ARTIFACTS            string
+		CAPM3_REL_TO_VERSION string
+	}
+	home, err := os.UserHomeDir()
+	Expect(err).To(BeNil())
+	templateVars := vars{artifactFolder, e2eConfig.GetVariable("CAPM3_REL_TO_VERSION")}
+
+	clusterctlUpgradeTestTemplate := filepath.Join(pwd, "/../../templates/test/upgrade/clusterctl-upgrade-test.yaml")
+	templates := template.Must(template.New("clusterctl-upgrade-test.yaml").ParseFiles(clusterctlUpgradeTestTemplate))
+	// Create a file to store the template output
+	clusterctlFile, err := os.Create(fmt.Sprintf("%s/.cluster-api/clusterctl.yaml", home))
+	Expect(err).To(BeNil())
+	// Execute the templates and write the output to the created file
+	err = templates.Execute(clusterctlFile, templateVars)
+	Expect(err).To(BeNil())
+	clusterctlFile.Close()
+
+	Logf("Upgrade management cluster to : %v", clusterv1.GroupVersion.Version)
+	clusterctl.UpgradeManagementClusterAndWait(ctx, clusterctl.UpgradeManagementClusterAndWaitInput{
+		ClusterctlConfigPath: e2eConfig.GetVariable("CONFIG_FILE_PATH"),
+		ClusterProxy:         upgradeClusterProxy,
+		Contract:             clusterv1.GroupVersion.Version,
+		LogFolder:            filepath.Join(artifactFolder, "clusters", workloadClusterName),
+	}, e2eConfig.GetIntervals(specName, "wait-controllers")...)
+
+	By("THE MANAGEMENT CLUSTER WAS SUCCESSFULLY UPGRADED!")
+
+	By("Get the management cluster images after upgrading")
+	printImages(upgradeClusterProxy)
+
+	By("Upgrade bootstrap cluster")
+	By("Get the management cluster images before upgrading")
+	printImages(bootstrapClusterProxy)
+
+	Logf("Upgrade bootstrap cluster to : %v", clusterv1.GroupVersion.Version)
+	clusterctl.UpgradeManagementClusterAndWait(ctx, clusterctl.UpgradeManagementClusterAndWaitInput{
+		ClusterctlConfigPath: e2eConfig.GetVariable("CONFIG_FILE_PATH"),
+		ClusterProxy:         bootstrapClusterProxy,
+		Contract:             clusterv1.GroupVersion.Version,
+		LogFolder:            filepath.Join(artifactFolder, "clusters", bootstrapClusterProxy.GetName()),
+	}, e2eConfig.GetIntervals(specName, "wait-controllers")...)
+	By("Get the bootstrap cluster images after upgrading")
+	printImages(bootstrapClusterProxy)
+
+	By("UPGRADE MANAGEMENT CLUSTER PASSED!")
+}
+
+func downloadToTmpFile(url string) string {
+	tmpFile, err := ioutil.TempFile("", "clusterctl")
+	Expect(err).ToNot(HaveOccurred(), "failed to get temporary file")
+	defer tmpFile.Close()
+
+	// Get the data
+	resp, err := http.Get(url) //nolint:gosec
+	Expect(err).ToNot(HaveOccurred(), "failed to get clusterctl")
+	defer resp.Body.Close()
+
+	// Write the body to file
+	_, err = io.Copy(tmpFile, resp.Body)
+	Expect(err).ToNot(HaveOccurred(), "failed to write temporary file")
+
+	return tmpFile.Name()
+}
```

Comment:
We could reuse the already existing `downloadFile` [func](https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/test/e2e/common.go#L70-L89) from the common.go